### PR TITLE
Cmake based build system

### LINKED ISSUE #2654: Call for Help: Proper Build System (CMake, Bazel, etc).
Currently vLLM's compilation tool uses PyTorch's extension builders, which calls Ninja under the hood. This works okay but have the following issues:
* Only supports NVIDIA and AMD GPUs. 
* Slow sequential builds. This is amplified by adding quantization kernels and LoRA kernels. 
* No caching and incremental builds. 

We would liked to ask for community's help on recommending a technology, prototype, and implement it. Ideally something like CMake or Bazel could work but it requires some careful thinking. 

The requirements:
* Must support multiple hardware architecture (NVIDIA, AMD, Intel, etc). 
* Must support incremental build, which also implies caching. 
* Must support parallelizable build.
* Good to have editor support (by generating compilation database). 
* Ideally it would not OOM like current setup. Currently due to the rigid structure, we have to carefully set `MAX_JOBS` and `NVCC_THREADS` to get around compiler goes out of memory. I think this is because nvcc spawn threads for each SM architecture we are compiling to. 
* vaguely, "future proof". 

Currently, the "build system" is all in here https://github.com/vllm-project/vllm/blob/main/setup.py

### PR DESCRIPTION
**Pull request by bnellnm**
This PR replaces the setup.py extension build system with a cmake build system.  I've tried to preserve and move all the existing logic about flags and supported architectures into the CMakeLists.txt file so that it is standalone, i.e. you should be able to build the C++/CUDA artifacts without going thru python or setup.py.  All the new setup.py does is select which extensions to build and the threading level for the nvcc compiler.  The cmake build should also honor most of the same environment variables as setup.py.  Some notable exceptions are `NVCC_THREADS` and `VLLM_ENABLE_PUNICA_KERNELS` which are replaced by command line cmake defines, e.g. `cmake ... -DNVCC_THREADS=8`.

The CMakeLists.txt defines three targets:`_C`, `_moe_C` and `_punica_C` which correspond to the names of the object files that were previously generated by the old setup.py.

There's also a `default` target which triggers the proper combination of `_C`, `_moe_C` and `_punica_C` based on the platform, architecture and `VLLM_ENABLE_PUNICA_KERNELS` variable.  This makes standalone use of cmake easier since the user doesn't need to know which specific extension targets need to be built for the current platform, e.g.
```
mkdir build && cd build
cmake -G Ninja -DVLLM_PYTHON_EXECUTABLE=$(which python3) -DVLLM_INSTALL_PUNICA_KERNELS=1 -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=../vllm ..
cmake --build . --target default  # will build _C, _moe_C and _punica_C if appropriate.
```
The CMakeLists.txt logic is a duplicate of the logic in setup.py that is used to select the extensions to build and should be kept in sync.

Some notable cmake variables for controlling versions:
- `PYTHON_SUPPORTED_VERSIONS` - lists the python versions supported by vllm.  If a new version of python is needed (or an old one removed), this variable should be updated.
- `CUDA_SUPPORTED_ARCHS` - a semi-colon separated list (i.e. a cmake list) of supported NVIDIA architectures. (this used to be NVIDIA_SUPPORTED_ARCHS in setup.py)
- `HIP_SUPPORTED_ARCHS` - a cmake list of supported AMD/ROCM architectures. (this used to be ROCM_SUPPORTED_ARCHS in setup.py)

The source files for each extension are controlled by the following variables which must be updated if the list of files changes (note that globs are not used since this is not recommended practice for cmake):
- `VLLM_EXT_SRC` for the _C extension.
- `VLLM_MOE_EXT_SRC` for the _moe_C extension.
- `VLLM_PUNICA_EXT_SRC` for the _punica_C extension.

Incremental builds should be supported "out of the box" but `ccache` will also be used if it is present.  The build can also be done with traditional makefiles if ninja isn't installed.  I've tried to balance the build parallelism with number of nvcc threads but this is a heuristic that might need to be tweaked.

The build flavor is controlled by the `CMAKE_BUILD_TYPE` environment variable with the default being `RelWithDebInfo`.  To change the default, edit the `cmake_build_ext` class in `setup.py`.

None of the build systems I looked at (cmake, meson, bazel) had any specific support for other GPU architectures.  Although the popularity of cmake probably makes it most likely to get support for new platforms.

I've tried to comment the CMakeLists.txt file with a description of the configuration process.  There were a few quirky bits that needed special handling.  In particular, setting the CUDA architecture flags required a somewhat convoluted process due to how torch/cmake interact.

For a ROCm build, there is an additional "hipify" preprocessor step that is run on the CUDA source files.  The `hipify.py` script is a command line wrapper around pytorch's hipify function that is used by `CMakeLists.txt`.  For uniformity, I've hooked the preprocessor up for all targets (not just `_C`) even though `_moe_C` and `_punica_C` currently aren't compiled (or supported?) for ROCm.

resolves https://github.com/vllm-project/vllm/issues/2654

### CODE REVIEW

**Comment by lroberts7** [comment]
@bnellnm is there a build command for nvidia gpu with this new setup? 

I'd like to try the build on our a100 to confirm I can still build once this is merged so I can continue to test and provide comments on open PRs. 

I'm currently able to build on tip of main. Do I need to set any specific environment variables or should something like the usual: 

```bash
mdkir build && cd build
cmake ..
```
work?

**Comment by bnellnm** [comment]
> @bnellnm is there a build command for nvidia gpu with this new setup?
> 
> I'd like to try the build on our a100 to confirm I can still build once this is merged so I can continue to test and provide comments on open PRs.
> 
> I'm currently able to build on tip of main. Do I need to set any specific environment variables or should something like the usual:
> 
> ```shell
> mdkir build && cd build
> cmake ..
> ```
> 
> work?

You should be able to use cmake directly w/o going thru python.   The cmake build should honor the same pytorch architecture environment variables, e.g. `TORCH_CUDA_ARCH_LIST`.  I just verified that the following works:
```
mkdir build && cd build
cmake -G Ninja ..
cmake --build . --target _C
```
The only caveat is that cmake will try to build all the extensions (`_C`, `_moe_C`, `_punica_C`) by default, so if you only want to build one, e.g. `_C` you'll have to specify that as a target.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+  macro(string_to_ver OUT_VER IN_STR)
+    string(REGEX REPLACE "\([0-9]+\)\([0-9]\)" "\\1.\\2" ${OUT_VER} ${IN_STR})
+  endmacro()
+
+  # Initialize the architecture lists to empty.
+  set(VLLM_GPU_ARCHES)
+  set(VLLM_PUNICA_GPU_ARCHES)
+
+  # Process each `gencode` flag.
+  foreach(ARCH ${VLLM_CUDA_ARCH_FLAGS})
```
Comment (simon-mo) [inline CMakeLists.txt:288]: thank you so much for supporting the existing workflow of `VLLM_CUDA_ARCH_FLAGS`. greatly appreciated!

**Review by simon-mo — COMMENTED** [review]
should we rename `supported` to `default`?

**Comment by simon-mo** [comment]
This looks great to me overall. @pcmoritz please do a detail pass.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+
+  if(NOT PYTHON_ERROR_CODE EQUAL 0)
+    message(FATAL_ERROR "${ERR_MSG}: ${PYTHON_STDERR}")
+  endif()
+endmacro()
+
+#
+# Try to find the MPI package.
+#
+find_package(MPI)
```
[the remark this answers is not in the document]
Comment (bnellnm) [reply CMakeLists.txt:89]: I found an inclusion of mpi.h here:
```
./csrc/custom_all_reduce_test.cu:24:#include "mpi.h"
```

File: setup.py [hunk 3]
```diff
@@ -1,31 +1,119 @@
...
-
 MAIN_CUDA_VERSION = "12.1"
 
-# Supported NVIDIA GPU architectures.
-NVIDIA_SUPPORTED_ARCHS = {"7.0", "7.5", "8.0", "8.6", "8.9", "9.0"}
-ROCM_SUPPORTED_ARCHS = {"gfx908", "gfx90a", "gfx942", "gfx1100"}
-# SUPPORTED_ARCHS = NVIDIA_SUPPORTED_ARCHS.union(ROCM_SUPPORTED_ARCHS)
+
+def is_ccache_available() -> bool:
+    return which("ccache") is not None
```
[the remark this answers is not in the document]
Comment (bnellnm) [reply setup.py:20]: I'll add a check for `sccache` before `ccache`

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (bnellnm) [reply CMakeLists.txt:89]: It looks like this is just test code that isn't normally included.  I think I ran into it because I was using a file glob to pick up all the .cu files at one point.  Thanks for pointing this out, I'll remove it.

[the remark this answers is not in the document]
Comment (hanzhi713) [reply CMakeLists.txt]: It does require MPI, but it's not needed to run vllm. It's a standalone test that may be used in development.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+# Run `EXPR` in python after importing `PKG`. Use the result of this to extend
+# `CMAKE_PREFIX_PATH` so the torch cmake configuration can be imported.
+macro (append_cmake_prefix_path PKG EXPR)
+  run_python(PREFIX_PATH
+    "import ${PKG}; print(${EXPR})" "Failed to locate ${PKG} path")
+  list(APPEND CMAKE_PREFIX_PATH ${PREFIX_PATH})
+endmacro()
+
+# Add user site-packages path to `CMAKE_PREFIX_PATH`.
+append_cmake_prefix_path("site" "site.getusersitepackages()")
```
Comment (pcmoritz) [inline CMakeLists.txt:99]: Why is this needed / done?

Comment (bnellnm) [reply CMakeLists.txt]: this was leftover cruft that isn't needed anymore.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+set(TORCH_SUPPORTED_VERSION_ROCM_6X "2.1.1")
+
+#
+# Loop through all supported python versions until the first suitable version
+# that has torch installed.
+#
+# Note: cmake is unable to pick the lowest supported version when multiple
+# versions are available, even with `CMAKE_FIND_PACKAGE_SORT_ORDER`.
+#
+foreach(VER ${PYTHON_SUPPORTED_VERSIONS})
```
[the remark this answers is not in the document]
Comment (bnellnm) [reply CMakeLists.txt]: I've changed the python find process to use the python binary path as specified by `VLLM_PYTHON_EXECUTABLE`.  Cmake will error out if this variable isn't set at config time and will also error if the specified python is not one of the supported versions.

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+
+#
+# Try to find python package with an executable that exactly matches
+# `VLLM_PYTHON_EXECUTABLE` and is one of the supported versions.
+#
+if (VLLM_PYTHON_EXECUTABLE)
+  find_python_from_executable(${VLLM_PYTHON_EXECUTABLE} "${PYTHON_SUPPORTED_VERSIONS}")
+else()
+  message(FATAL_ERROR
+    "Please set VLLM_PYTHON_EXECUTABLE to the desired python version before "
```
Comment (pcmoritz) [inline CMakeLists.txt:41]: `to the path of the desired python interpreter`

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+# Add the `default` target which detects which extensions should be
+# built based on platform/architecture.  This is the same logic that
+# setup.py uses to select which extensions should be built and should
+# be kept in sync.
+#
+# The `default` target makes direct use of cmake easier since knowledge
+# of which extensions are supported has been factored in, e.g.
+#
+# mkdir build && cd build
+# cmake -G Ninja ..
```
Comment (pcmoritz) [inline CMakeLists.txt:381]: Here we should say
```
cmake -G Ninja -DVLLM_PYTHON_EXECUTABLE=`which python` ..
```

File: Dockerfile [hunk 8]
```diff
@@ -38,6 +38,9 @@ RUN --mount=type=cache,target=/root/.cache/pip \
 # copy input files
 COPY csrc csrc
 COPY setup.py setup.py
+COPY hipify.py hipify.py
```
Comment (pcmoritz) [inline Dockerfile:41]: I think the `hipify.py` and `utils.cmake` files should go into a `cmake` folder to avoid having too many files in the root directory.

File: CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+# Torch also imports CUDA (and partially HIP) languages with some customizations,
+# so there is no need to do this explicitly with check_language/enable_language,
+# etc.
+#
+find_package(Torch REQUIRED)
+
+# For some reason torch does not add libtorch_python.so to the list of torch
+# libraries to link.  Find it by hand using `append_torchlib_if_found` from
+# torch's cmake setup.
+append_torchlib_if_found(torch_python)
```
Comment (pcmoritz) [inline CMakeLists.txt:63]: Have you looked more into this? It is very odd that this is needed :) What are the libraries that are currently included in `${TORCH_LIBRARIES}`? Which symbols are missing from them?

Comment (bnellnm) [reply CMakeLists.txt:63]: The first undefined symbol I encountered was `pybind11::detail::type_caster<at::Tensor, void>::load(pybind11::handle, bool)` but there may be more.

This symbol is defined in `libtorch_python.so` which would normally have been added by `torch.utils.cpp_extension.CUDAExtension` if setup.py was using torch's extension mechanism to build.

The cmake config from `find_pacakge(Torch)` does not add this library since I think it assumes it is only going to be used for standalone C++ binaries that link against torch.   See `torch/share/cmake/Torch/TorchConfig.cmake` where it adds all the other libraries.

Comment (bnellnm) [reply CMakeLists.txt:63]: I'll update the comment to include the details.

**Comment by bnellnm** [comment]
@pcmoritz, @simon-mo I had a question about the punica kernels setup.  It looks like the existing `setup.py` is checking the physical CUDA devices to see if they supported (>= 8.0) rather than just depending on whatever `TORCH_CUDA_ARCH_LIST` was set to.  This seems wrong if we are cross compiling?  It seems to me that checking `VLLM_INSTALL_PUNICA_KERNELS` and making sure that the target arch list for punica only contains arches >= 8.0 would be more correct?

I was trying to replicate the logic from `setup.py` in the `CMakeFile.txt` and was having some trouble and noticed that in the buildkite docker there don't seem to be any GPUs, e.g. from a recent log
```
#24 5.315 -- CUDA supported arches: 7.0;7.5;8.0;8.6;8.9;9.0
#24 5.315 -- CUDA target arches: 70;75;80;86;89;90;90-virtual
#24 7.002 -- Punica target arches: 80;86;89;90;90-virtual
#24 7.002 -- Enabling C extension.
#24 7.002 -- Enabling moe extension.
#24 7.002 -- arches: OFF
#24 7.002 -- native arches: No CUDA devices found.-real
#24 7.002 -- ARCH_VER_STR:
#24 7.002 -- Unable to add punica extension due to device version  < 8.0.
#24 7.002 CMake Error at CMakeLists.txt:275 (string_to_ver):
#24 7.002   string_to_ver Macro invoked with incorrect arguments for macro named:
#24 7.002   string_to_ver
```

**Comment by bnellnm** [comment]
> @bnellnm Thanks for all the updates, I have one more questions: Have you looked into whether the hippification can be done without duplicating the logic from pytorch? How does the pytorch based CMake build system do it? If we can do it the same way without poking through the abstractions, it will very likely be more maintainable going forward.
> 
> On the punica question, I don't exactly know -- your logic seems simpler. I think we need to make sure that these two things keep working:
> 
> * Punica will work for the architectures that it is supported on for the release wheels
> * For developers / people who compile vllm themselves, if they switch on the `VLLM_INSTALL_PUNICA_KERNELS` flag and their hardware supports punica, it will just work

@pcmoritz Thanks for all the good feedback.

The `hipify.py` script I added is CLI wrapper around the "hipification" process used by pytorch's extension system.  When building through python it is invoked from `torch.utils.cpp_extension.CUDAExtension` directly and not through cmake.  The code in pytorch appears to be a copy (or modified copy) of `https://github.com/ROCm/hipify_torch`.  I decided to go thru pytorch's version of hipify since I didn't want to add an extra external dependency and I didn't know if there had been any customization or special dependency on the existing version of the code.  

As for the punica kernels, I'm pretty sure they will work as expected since the build now follows the same process as the other kernels, with the exception of filtering out additional unsupported architectures.  And manually setting `TORCH_CUDA_ARCH_LIST` will allow compiling the punica kernels no matter what the version is for the underlying hardware.

**Comment by pcmoritz** [comment]
Here are some benchmarks -- these all look very reasonable:

On this branch, `time pip install -e .` takes
```
real    3m55.800s
user    11m1.568s
sys     1m15.713s
```

On master, `time pip install -e .` takes
```
real    3m33.912s
user    10m48.451s
sys     0m54.333s
```

Incremental compilation on this branch `touch csrc/cache_kernels.cu && time python setup.py develop`:
```
real    0m58.882s
user    0m57.193s
sys     0m27.270s
```
and ncremental compilation with cmake: `touch csrc/cache_kernels.cu && time cmake --build . --target default`
```
real    0m37.769s
user    0m34.976s
sys     0m2.763s
```

On master, `touch csrc/cache_kernels.cu && time python setup.py develop`:
```
real    0m46.619s
user    0m42.561s
sys     0m12.813s
```

File: setup.py [hunk 10]
```diff
@@ -45,27 +180,26 @@ def _is_cuda() -> bool:
     return (torch.version.cuda is not None) and not _is_neuron()
 
 
-# Compiler flags.
-CXX_FLAGS = ["-g", "-O2", "-std=c++17"]
-# TODO(woosuk): Should we use -O3?
-NVCC_FLAGS = ["-O2", "-std=c++17"]
+def _install_punica() -> bool:
+    print(f"DEBUG device_count = {torch.cuda.device_count()}")
```
Comment (pcmoritz) [inline setup.py:184]: Can you remove this?

**Comment by pcmoritz** [comment]
@bnellnm I finished reviewing this and making sure that the binary is working (and also that punica is working if VLLM_INSTALL_PUNICA_KERNELS=1 is set during the build). Great work!

If we merge this PR, are you willing to maintain the CMake files going forward? We will probably need to add a decent amount more kernels to the build going forward, including https://github.com/vllm-project/vllm/pull/3221 :)

**Review by pcmoritz — APPROVED** [review]

**Comment by simon-mo** [comment]
I'll try to merge #2886 by EOD today, and you can run sanity test this on AMD CI

**Review by simon-mo — APPROVED** [review]

**Comment by hliuca** [comment]
The changes in setup.py messed up HIP support.

**Comment by simon-mo** [comment]
Hi @hliuca, this PR compiled on AMD machine. Can you elaborate on the issue?

**Comment by hliuca** [comment]
hon-310/vllm/model_executor/layers/fused_moe/configs
running build_ext
Traceback (most recent call last):
  File "/workdir/vllm-mlperf/setup.py", line 338, in <module>
    setup(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/__init__.py", line 103, in setup
    return distutils.core.setup(**attrs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 185, in setup
    return run_commands(dist)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
    dist.run_commands()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
    self.run_command(cmd)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/dist.py", line 963, in run_command
    super().run_command(command)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
    cmd_obj.run()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/command/install.py", line 85, in run
    self.do_egg_install()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/command/install.py", line 137, in do_egg_install
    self.run_command('bdist_egg')
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
    self.distribution.run_command(command)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/dist.py", line 963, in run_command
    super().run_command(command)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
    cmd_obj.run()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/command/bdist_egg.py", line 167, in run
    cmd = self.call_command('install_lib', warn_dir=0)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/command/bdist_egg.py", line 153, in call_command
    self.run_command(cmdname)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
    self.distribution.run_command(command)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/dist.py", line 963, in run_command
    super().run_command(command)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
    cmd_obj.run()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/command/install_lib.py", line 11, in run
    self.build()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/command/install_lib.py", line 111, in build
    self.run_command('build_ext')
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
    self.distribution.run_command(command)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/dist.py", line 963, in run_command
    super().run_command(command)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
    cmd_obj.run()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/command/build_ext.py", line 89, in run
    _build_ext.run(self)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 345, in run
    self.build_extensions()
  File "/workdir/vllm-mlperf/setup.py", line 157, in build_extensions
    self.configure(ext)
  File "/workdir/vllm-mlperf/setup.py", line 125, in configure
    num_jobs, nvcc_threads = self.compute_num_jobs()
  File "/workdir/vllm-mlperf/setup.py", line 64, in compute_num_jobs
    nvcc_cuda_version = get_nvcc_cuda_version()
  File "/workdir/vllm-mlperf/setup.py", line 237, in get_nvcc_cuda_version
    nvcc_output = subprocess.check_output([CUDA_HOME + "/bin/nvcc", "-V"],
TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'

**Comment by hliuca** [comment]
def _is_cuda() -> bool:
    return (torch.version.cuda is not None) and not _is_neuron()

was changed to
def _is_cuda() -> bool:
    return torch.version.cuda is not None

**Comment by hliuca** [comment]
the following function checks nvcc, and it doesn't seem to check HIP.

class cmake_build_ext(build_ext):
    # A dict of extension directories that have been configured.
    did_config = {}

    #
    # Determine number of compilation jobs and optionally nvcc compile threads.
    #
    def compute_num_jobs(self):
        try:
            # os.sched_getaffinity() isn't universally available, so fall back
            # to os.cpu_count() if we get an error here.
            num_jobs = len(os.sched_getaffinity(0))
        except AttributeError:
            num_jobs = os.cpu_count()

        nvcc_cuda_version = get_nvcc_cuda_version()
        if nvcc_cuda_version >= Version("11.2"):
            nvcc_threads = int(os.getenv("NVCC_THREADS", 8))
            num_jobs = max(1, round(num_jobs / (nvcc_threads / 4)))
        else:
            nvcc_threads = None

        return num_jobs, nvcc_threads

**Comment by bnellnm** [comment]
> the following function checks nvcc, and it doesn't seem to check HIP.
> 
> class cmake_build_ext(build_ext): # A dict of extension directories that have been configured. did_config = {}
> 
> ```
> #
> # Determine number of compilation jobs and optionally nvcc compile threads.
> #
> def compute_num_jobs(self):
>     try:
>         # os.sched_getaffinity() isn't universally available, so fall back
>         # to os.cpu_count() if we get an error here.
>         num_jobs = len(os.sched_getaffinity(0))
>     except AttributeError:
>         num_jobs = os.cpu_count()
> 
>     nvcc_cuda_version = get_nvcc_cuda_version()
>     if nvcc_cuda_version >= Version("11.2"):
>         nvcc_threads = int(os.getenv("NVCC_THREADS", 8))
>         num_jobs = max(1, round(num_jobs / (nvcc_threads / 4)))
>     else:
>         nvcc_threads = None
> 
>     return num_jobs, nvcc_threads
> ```

I think a check for `_is_cuda()` in `compute_num_jobs` should fix the problem.

**Comment by hliuca** [comment]
> > the following function checks nvcc, and it doesn't seem to check HIP.
> > class cmake_build_ext(build_ext): # A dict of extension directories that have been configured. did_config = {}
> > ```
> > #
> > # Determine number of compilation jobs and optionally nvcc compile threads.
> > #
> > def compute_num_jobs(self):
> >     try:
> >         # os.sched_getaffinity() isn't universally available, so fall back
> >         # to os.cpu_count() if we get an error here.
> >         num_jobs = len(os.sched_getaffinity(0))
> >     except AttributeError:
> >         num_jobs = os.cpu_count()
> > 
> >     nvcc_cuda_version = get_nvcc_cuda_version()
> >     if nvcc_cuda_version >= Version("11.2"):
> >         nvcc_threads = int(os.getenv("NVCC_THREADS", 8))
> >         num_jobs = max(1, round(num_jobs / (nvcc_threads / 4)))
> >     else:
> >         nvcc_threads = None
> > 
> >     return num_jobs, nvcc_threads
> > ```
> 
> I think a check for `_is_cuda()` in `compute_num_jobs` should fix the problem.

yes. also, _is_cuda should be restored... new version doesn't work :-)

def _is_cuda() -> bool:
    return (torch.version.cuda is not None) and not _is_neuron()

**Comment by bnellnm** [comment]
> > > the following function checks nvcc, and it doesn't seem to check HIP.
> > > class cmake_build_ext(build_ext): # A dict of extension directories that have been configured. did_config = {}
> > > ```
> > > #
> > > # Determine number of compilation jobs and optionally nvcc compile threads.
> > > #
> > > def compute_num_jobs(self):
> > >     try:
> > >         # os.sched_getaffinity() isn't universally available, so fall back
> > >         # to os.cpu_count() if we get an error here.
> > >         num_jobs = len(os.sched_getaffinity(0))
> > >     except AttributeError:
> > >         num_jobs = os.cpu_count()
> > > 
> > >     nvcc_cuda_version = get_nvcc_cuda_version()
> > >     if nvcc_cuda_version >= Version("11.2"):
> > >         nvcc_threads = int(os.getenv("NVCC_THREADS", 8))
> > >         num_jobs = max(1, round(num_jobs / (nvcc_threads / 4)))
> > >     else:
> > >         nvcc_threads = None
> > > 
> > >     return num_jobs, nvcc_threads
> > > ```
> > 
> > 
> > I think a check for `_is_cuda()` in `compute_num_jobs` should fix the problem.
> 
> yes. also, _is_cuda should be restored... new version doesn't work :-)
> 
> def _is_cuda() -> bool: return (torch.version.cuda is not None) and not _is_neuron()

This PR didn't change the definition of `_is_cuda()`, that was from a prior PR #2671

**Comment by bnellnm** [comment]
I think this PR will fix the problem #3481

**Comment by hliuca** [comment]
is_cuda returns True, even it is used in HIP. The change in #2671 was not correct :-)
